### PR TITLE
8290200: com/sun/jdi/InvokeHangTest.java fails with "Debuggee appears to be hung"

### PR DESCRIPTION
**Pull request by plummercj**
The debuggee main method creates two threads and then starts them:

```
    public static void main(String[] args) {
        System.out.println("Howdy!");
        Thread t1 = TestScaffold.newThread(new InvokeHangTarg(), name1);
        Thread t2 = TestScaffold.newThread(new InvokeHangTarg(), name2);

        t1.start();
        t2.start();
    }
```

These threads will hit breakpoints which the debugger handles and issues an invoke on the breakpoint thread. The threads run until they generate 100 breakpoints. There is an issue when these two threads are virtual threads. Virtual threads are daemon threads. That means the JVM can exit while they are still running. The above main() method is not waiting for these two threads to exit, so main() exits immediately and the JVM starts the shutdown process. It first must wait for all non-daemon threads to exit, but there are none, so the JVM exits right away before the two threads are completed.  The end result of this early exit is that sometimes the invoke done by the debugger never completes because the JVM has already issued a VMDeath event and the debuggee has been disconnected.

When these two threads are platform threads, the JVM has to wait until they complete before it exits, so they will always complete. The fix for virtual threads is to do a join with t1 and t2. This forces the main() method to block until they have completed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290200](https://bugs.openjdk.org/browse/JDK-8290200): com/sun/jdi/InvokeHangTest.java fails with "Debuggee appears to be hung"

### Reviewers
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**) ⚠️ Review applies to [4d28b92e](https://git.openjdk.org/jdk/pull/13068/files/4d28b92e413f21000cf3bc392b8907527d496722)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [4d28b92e](https://git.openjdk.org/jdk/pull/13068/files/4d28b92e413f21000cf3bc392b8907527d496722)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [4d28b92e](https://git.openjdk.org/jdk/pull/13068/files/4d28b92e413f21000cf3bc392b8907527d496722)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [58ca27b8](https://git.openjdk.org/jdk/pull/13068/files/58ca27b8e590cac9311f1998ad560c2032e41194)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/13068/head:pull/13068` \
`$ git checkout pull/13068`

Update a local copy of the PR: \
`$ git checkout pull/13068` \
`$ git pull https://git.openjdk.org/jdk pull/13068/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13068`

View PR using the GUI difftool: \
`$ git pr show -t 13068`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13068.diff">https://git.openjdk.org/jdk/pull/13068.diff</a>

</details>

### CODE REVIEW

File: test/jdk/com/sun/jdi/InvokeHangTest.java [hunk 1]
```diff
@@ -214,7 +220,7 @@ protected void runTests() throws Exception {
         targetClass = bpe.location().declaringType();
         mainThread = bpe.thread();
         EventRequestManager erm = vm().eventRequestManager();
-        final Thread mainThread = Thread.currentThread();
+        final Thread mainTestThread = Thread.currentThread();
```
Comment (plummercj) [inline test/jdk/com/sun/jdi/InvokeHangTest.java:223]: This local variable was conflicting with an instance field of the same name. See line 135. I ran into issues when I added some debugging code to this method that referenced the other `mainThread`, so I did a rename.

File: test/jdk/com/sun/jdi/InvokeHangTest.java [hunk 2]
```diff
@@ -191,7 +197,7 @@ public void breakpointReached(BreakpointEvent event) {
         ThreadReference thread = event.thread();
         try {
             StackFrame sf = thread.frame(0);
-            System.err.println("  Debugger: Breakpoint hit at "+sf.location());
+            System.out.println("  Debugger: Breakpoint hit at "+sf.location());
```
Comment (alexmenkov) [inline test/jdk/com/sun/jdi/InvokeHangTest.java:200]: while you are here please add spaces around plus

**Review by alexmenkov — APPROVED** [review]

**Review by lmesnik — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

File: test/jdk/com/sun/jdi/InvokeHangTest.java [hunk 3]
```diff
@@ -59,6 +59,12 @@ public static void main(String[] args) {
 
         t1.start();
         t2.start();
+        try {
+            t1.join();
+            t2.join();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (dcubed-ojdk) [inline test/jdk/com/sun/jdi/InvokeHangTest.java:67]: Please consider adding a comment that explains that the `join()` calls
are only necessary when `t1` and `t2` are virtual threads.

**Review by dcubed-ojdk — APPROVED** [review]

File: test/jdk/com/sun/jdi/InvokeHangTest.java [hunk 4]
```diff
@@ -59,11 +59,23 @@ public static void main(String[] args) {
 
         t1.start();
         t2.start();
+
+        try {
+            // The join ensures that the test completes before we exit main(). If we are using
+            // virtual threads, they are always daemon threads, and therefor the JVM will exit
```
Comment (dcubed-ojdk) [inline test/jdk/com/sun/jdi/InvokeHangTest.java:65]: nit typo: s/therefor/therefore/

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. Thanks for adding the comment.
One possible nit typo...

**Pull request closed without merging**